File: modules/backup-storage/ios/BackupStorageModule.swift
import CryptoKit
import Darwin
import ExpoModulesCore
import Foundation

private let backupProcessId = UUID().uuidString.lowercased()
private let controlLock = NSLock()

public class BackupStorageModule: Module {
  public func definition() -> ModuleDefinition {
    Name("BackupStorage")
    Function("processId") { backupProcessId }
    Function("readControl") { (documentUri: String) throws -> String? in
      controlLock.lock()
      defer { controlLock.unlock() }
      let file = try self.controlFile(documentUri)
      guard FileManager.default.fileExists(atPath: file.path) else { return nil }
      let data = try Data(contentsOf: file)
      guard data.count <= 16384, let value = String(data: data, encoding: .utf8) else {
        throw self.failure("Invalid storage control record")
      }
      return value
    }
    Function("writeControl") { (documentUri: String, value: String) throws in
      controlLock.lock()
      defer { controlLock.unlock() }
      guard let bytes = value.data(using: .utf8), bytes.count <= 16384 else {
        throw self.failure("Invalid storage control record")
      }
      let file = try self.controlFile(documentUri)
      let parent = file.deletingLastPathComponent()
      try FileManager.default.createDirectory(at: parent, withIntermediateDirectories: true)
      let temporary = parent.appendingPathComponent("state.tmp")
      try bytes.write(to: temporary, options: .completeFileProtectionUntilFirstUserAuthentication)
      try self.flush(temporary)
      guard rename(temporary.path, file.path) == 0 else {
        throw self.failure("Could not replace storage control record")
      }
      try self.flush(parent)
      try self.flush(parent.deletingLastPathComponent())
    }
    AsyncFunction("hashFile") { (uri: String) throws -> String in
      let file = try self.privateFile(uri)
      let handle = try FileHandle(forReadingFrom: file)
      defer { try? handle.close() }
      var digest = SHA256()
      while let chunk = try handle.read(upToCount: 262144), !chunk.isEmpty {
        digest.update(data: chunk)
      }
      return digest.finalize().map { String(format: "%02x", $0) }.joined()
    }
    AsyncFunction("sealDirectory") { (uri: String) throws in
      let root = try self.privateFile(uri)
      try self.seal(root)
      try self.flush(root.deletingLastPathComponent())
    }
  }

  private func controlFile(_ documentUri: String) throws -> URL {
    try privateFile(documentUri).appendingPathComponent("storage-control/state.json")
  }

  private func privateFile(_ uri: String) throws -> URL {
    guard let file = URL(string: uri), file.isFileURL else {
      throw failure("Expected a private file URL")
    }
    let resolved = file.standardizedFileURL.resolvingSymlinksInPath()
    let home = URL(fileURLWithPath: NSHomeDirectory()).resolvingSymlinksInPath().path + "/"
    guard resolved.path.hasPrefix(home) else {
      throw failure("Path is outside private storage or contains a symlink")
    }
    return resolved
  }

  private func seal(_ file: URL) throws {
    let values = try file.resourceValues(forKeys: [.isDirectoryKey, .isSymbolicLinkKey, .isRegularFileKey])
    guard values.isSymbolicLink != true else { throw failure("Symlinks are not supported") }
    if values.isDirectory == true {
      for child in try FileManager.default.contentsOfDirectory(at: file, includingPropertiesForKeys: nil) {
        try seal(child)
      }
    } else if values.isRegularFile != true {
      throw failure("Unsupported storage entry")
    }
    try flush(file)
  }

  private func flush(_ file: URL) throws {
    let descriptor = open(file.path, O_RDONLY)
    guard descriptor >= 0 else { throw failure("Could not open storage for synchronization") }
    defer { close(descriptor) }
    guard fsync(descriptor) == 0 else { throw failure("Could not synchronize storage") }
  }

  private func failure(_ message: String) -> NSError {
    NSError(domain: "BackupStorage", code: 1, userInfo: [NSLocalizedDescriptionKey: message])
  }
}

File: modules/backup-storage/android/src/main/java/expo/modules/backupstorage/BackupStorageModule.kt
package expo.modules.backupstorage

import android.content.Intent
import android.net.Uri
import android.os.Process
import android.system.Os
import android.system.OsConstants
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import java.security.MessageDigest
import java.util.UUID

private val backupProcessId = UUID.randomUUID().toString()
private val controlLock = Any()

class BackupStorageModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("BackupStorage")
    Function("processId") { backupProcessId }
    AsyncFunction("restartAfterRestore") {
      val context = requireNotNull(appContext.reactContext).applicationContext
      check(context.packageManager.getLaunchIntentForPackage(context.packageName) != null) {
        "Could not find the app launch activity"
      }
      context.startActivity(
        Intent(context, BackupRestartActivity::class.java)
          .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
          .putExtra(EXTRA_ORIGINAL_PROCESS_ID, Process.myPid())
      )
    }
    Function("readControl") { documentUri: String ->
      synchronized(controlLock) {
        val file = controlFile(documentUri)
        if (!file.exists()) null else {
          require(file.length() <= 16384) { "Invalid storage control record" }
          file.readText(Charsets.UTF_8)
        }
      }
    }
    Function("writeControl") { documentUri: String, value: String ->
      synchronized(controlLock) {
        val bytes = value.toByteArray(Charsets.UTF_8)
        require(bytes.size <= 16384) { "Invalid storage control record" }
        val file = controlFile(documentUri)
        val parent = requireNotNull(file.parentFile)
        check(parent.isDirectory || parent.mkdirs()) { "Could not create control directory" }
        val temporary = File(parent, "state.tmp")
        FileOutputStream(temporary).use { stream ->
          stream.write(bytes)
          stream.fd.sync()
        }
        Os.rename(temporary.path, file.path)
        flush(parent)
        flush(requireNotNull(parent.parentFile))
      }
    }
    AsyncFunction("hashFile") { uri: String ->
      val digest = MessageDigest.getInstance("SHA-256")
      privateFile(uri).inputStream().buffered().use { stream ->
        val bytes = ByteArray(262144)
        while (true) {
          val size = stream.read(bytes)
          if (size < 0) break
          digest.update(bytes, 0, size)
        }
      }
      digest.digest().joinToString("") { "%02x".format(it.toInt() and 0xff) }
    }
    AsyncFunction("sealDirectory") { uri: String ->
      val root = privateFile(uri)
      seal(root)
      flush(requireNotNull(root.parentFile))
    }
  }

  private fun controlFile(documentUri: String) = File(privateFile(documentUri), "storage-control/state.json")

  private fun privateFile(uri: String): File {
    val parsed = Uri.parse(uri)
    require(parsed.scheme == "file") { "Expected a private file URL" }
    val file = File(requireNotNull(parsed.path)).absoluteFile
    val context = requireNotNull(appContext.reactContext)
    val root = File(context.applicationInfo.dataDir).canonicalPath + File.separator
    require(file.canonicalPath.startsWith(root)) {
      "Path is outside private storage or contains a symlink"
    }
    return file.canonicalFile
  }

  private fun seal(file: File) {
    require(file.absolutePath == file.canonicalPath) { "Symlinks are not supported" }
    if (file.isDirectory) {
      val children = requireNotNull(file.listFiles()) { "Could not list storage directory" }
      children.forEach { seal(it) }
    } else {
      require(file.isFile) { "Unsupported storage entry" }
    }
    flush(file)
  }

  private fun flush(file: File) {
    val descriptor = Os.open(file.path, OsConstants.O_RDONLY, 0)
    try { Os.fsync(descriptor) } finally { Os.close(descriptor) }
  }
}
